File: Oxcart.hs
module Oxcart where

data Elem = Num Integer
          | Cont ([Elem] -> [Elem])

instance Show Elem where
    show (Num n)         = show n
    show (Cont k)        = "#k"

type Op = [Elem] -> ([Elem] -> [Elem]) -> [Elem]


nop :: Op
nop  xs k = k xs

push0 :: Op
push0 xs k = k ((Num 0):xs)

incr :: Op
incr ((Num x):xs) k = k (Num (x+1):xs)

dbl :: Op
dbl  ((Num x):xs) k = k (Num (x*2):xs)

save :: Op
save xs k = k (Cont k:xs)

rsr :: Op
rsr xs@((Cont k):_) _ = k xs



m [] = nop
m (x:xs) = (m' x) `composeCPS` (m xs)
    where
        m' '0' = push0
        m' '+' = incr
        m' 'X' = dbl
        m' '*' = save
        m' '$' = rsr
        composeCPS f g = \x k -> (f x (\s -> (g s k)))


run s = m s [] id

test = run "+XX"

File: Oxcart.js
function Push(x, xs) {
  var xs2 = xs.slice();
  xs2.push(x);
  return xs2;
}

function Top(xs) {
  return xs[xs.length - 1];
}

function Pop(xs) {
  var xs2 = xs.slice();
  xs2.pop();
  return xs2;
}

/* --- */

function nop(xs, k) {
  return k(xs);
}

function push0(xs, k) {
  return k(Push(0, xs));
}

function incr(xs, k) {
  var v = Top(xs);
  return k(Push(v + 1, Pop(xs)));
}

function dbl(xs, k) {
  var v = Top(xs);
  return k(Push(v * 2, Pop(xs)));
}

function save(xs, k) {
  return k(Push(k, xs));
}

function rsr(xs, k) {
  var j = Top(xs);
  return j(Pop(xs));
}

function swpk(xs, k) {
  var j = Top(xs);
  return j(Push(k, Pop(xs)));
}

/* --- */

function composeCPS(f, g) {
  return function(x, k) {
    return f(x, function(s) {
      return g(s, k);
    });
  }
}

function compile(s) {
  var f = nop;
  for (var i = 0; i < s.length; i++) {
    var c = s.charAt(i);
    switch(c) {
      case '0':
        f = composeCPS(f, push0);
        break;
      case '+':
        f = composeCPS(f, incr);
        break;
      case 'X':
        f = composeCPS(f, dbl);
        break;
      case '*':
        f = composeCPS(f, save);
        break;
      case '$':
        f = composeCPS(f, rsr);
        break;
      case '_':
        f = composeCPS(f, swpk);
        break;
    }
  }
  return f;
}

function out(x) {
  console.log(x);
}

function run(s) {
  compile(s)([], out);
}
